src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptor.cs

```diff
@@ -14,139 +14,43 @@ namespace Microsoft.AspNet.Razor.TagHelpers
-        /// The attribute names required for the tag helper to target the HTML tag.
-        /// </param>
-        /// <param name="allowedChildren">HTML element tag names targeted by tag helpers allowed as children.</param>
-        /// <param name="tagStructure">The expected tag structure.</param>
-        /// <param name="designTimeDescriptor">The <see cref="TagHelperDesignTimeDescriptor"/> that contains design
-        /// time information about the tag helper.</param>
-        public TagHelperDescriptor(
-            string prefix,
-            [NotNull] string tagName,
-            [NotNull] string typeName,
-            [NotNull] string assemblyName,
-            [NotNull] IEnumerable<TagHelperAttributeDescriptor> attributes,
-            [NotNull] IEnumerable<string> requiredAttributes,
-            IEnumerable<string> allowedChildren,
-            TagStructure tagStructure,
-            TagHelperDesignTimeDescriptor designTimeDescriptor)
-        {
-            Prefix = prefix ?? string.Empty;
-            TagName = tagName;
-            FullTagName = Prefix + TagName;
-            TypeName = typeName;
-            AssemblyName = assemblyName;
-            Attributes = new List<TagHelperAttributeDescriptor>(attributes);
-            RequiredAttributes = new List<string>(requiredAttributes);
-            TagStructure = tagStructure;
-            DesignTimeDescriptor = designTimeDescriptor;
-
-            if (allowedChildren != null)
-            {
-                AllowedChildren = new List<string>(allowedChildren);
-            }
-        }
-
-        /// <summary>
         /// Text used as a required prefix when matching HTML start and end tags in the Razor source to available
         /// tag helpers.
         /// </summary>
-        public string Prefix { get; }
+        public string Prefix { get; set; } = string.Empty;
 
         /// <summary>
         /// The tag name that the tag helper should target.
         /// </summary>
-        public string TagName { get; }
+        public string TagName { get; [param: NotNull] set; }
 
         /// <summary>
         /// The full tag name that is required for the tag helper to target an HTML element.
         /// </summary>
         /// <remarks>This is equivalent to <see cref="Prefix"/> and <see cref="TagName"/> concatenated.</remarks>
-        public string FullTagName { get; }
+        public string FullTagName
+        {
+            get
+            {
+                return Prefix + TagName;
+            }
+        }
 
         /// <summary>
         /// The full name of the tag helper class.
         /// </summary>
-        public string TypeName { get; }
+        public string TypeName { get; [param: NotNull] set; }
 
         /// <summary>
         /// The name of the assembly containing the tag helper class.
         /// </summary>
-        public string AssemblyName { get; }
+        public string AssemblyName { get; [param: NotNull] set; }
```

Comment:
Note that nothing guarantees `TagName`, `TypeName`, or `AssemblyName` are ever set. Fortunately these descriptors are originally created in TagHelperDescriptorFactory -- after checking source attribute values. But may be worth adding Debug.Assert() expressions checking these values in `TagHelperDescriptorResolver`.